### ISSUE TITLE
teach us about the binaries

### ISSUE THREAD

**Issue by wizza-smile** [SublimeCodeIntel/SublimeCodeIntel#451]
I tracked this issue down:
https://github.com/SublimeCodeIntel/SublimeCodeIntel/issues/356

it's a bug in the "_ielementtree.so", where it gets confused about the scopes.

@Kronuz could you provide a step by step instruction (which sources, patches, compiler settings) on how to compile the current version of this one particular file, so that I can attempt to fix it?

I am aware that the main issue with the binaries is multiple platforms.

**Comment by Kronuz** [comment]
Absolutely. The reason i did't pull the PR you sent is this, there is a second repository CodeIntelSources, which contains all the needed building pieces in the src directory) where I build the releases from. Ideally, all pull requests should go to one of those repositories there. Historically, Komodo Editor's source code (was OpenKomodo) was in one (huge) Subversion repository. I built a script (`svn-rebase.sh --svn`) to:

1) "pull" and upgrade the SVN base into several different repositories (all of them in src), _[svn branch]_

2) appliy all the Komodo provided patches. Only the ones for scintilla, `contrib/patches/scintilla`, as the others where already applied in the svn, for some reason AND the ones in `src/patches/ciElementTreee*`, which are used to build the ciElementTree out of the already patches cElementTree in `contrib` _[patched branch]_

3) do a PEP-8 normalization across the source code _[parched-pep8 branch]_

4) apply SublimeCodeIntel patches (master branch),

4) ran 2to3 over the code in master branches, for the ST3 _[py3 branch]_, and

5) applied additional patches for Python 3 compatibility missed by the 2to3 tool _[development branch]_

6) for building, I use the `build.sh` script on each platform and then the `deploy.sh` on both the `master` (ST2) and `development` (ST3) branches, for bundling everything together for a release in SublimeCodeIntel repository.

The particular iElementTree you are referring to, I extracted it from Python3's source code (ElementTree) and manually applied all patches from Komodo's cElementTree and ciElementTree (as making those last two ones work under Python3 proved close to impossible for me, even when I got them to compile). This one is also compiled during the building process when using `build.sh`.

https://github.com/SublimeCodeIntel/CodeIntelSources/blob/master/src/README.rst outlines the building process and at the bottom are the instructions about how to upgrade everything else (which is now changing, as they no longer have the SVN reposiyory . I'd be pleased to add further instructions or clarify anything else there if something is still confusing.

I'm currently working in a simplified workflow, now that they've moved to GitHub. I'm which I think I'll keep a KomodoEditor fork with the PEP-8 and patches applied, and start extracting commits from there as git patches into the CodeIntelSource repositories. We could discuss the better way if you wish.

**Comment by wizza-smile** [comment]
Sorry I didn't answer earlier, my brain had to cope with your answer ;)

Your internal workflow is irrelevant (to me and possible contributors).
What you should provide is a source file (like a cElementTree.c) and a setup.py on which I can execute a ```python setup.py build```and then get the current version of the binary for my platform.

These files could be provided as a repo to which people can add pull-requests.

Your job as a maintainer would be to keep an overview and ask for specific help where needed!

Please re-open this issue so people can find it!

**Comment by wizza-smile** [comment]
@Kronuz what are your thoughts on this?

**Comment by Kronuz** [comment]
I guess it's sort of ok. I already have repositories for each of the libraries... The thing with the workflow is just that I want to more easily be able to pull changes from the original Komodo Editor, as they maintain the codebase; hence the whole mumbo-jumbo about the branches... and as much as I wish I was immortal, I'm not, and if I'm ever not around, I'd want for someone else to know what my workflow was, so anybody can continue from there. In that sense, I do not concur that the workflows of an open source project are irrelevant.

**Comment by wizza-smile** [comment]
The way it is now, it is not immediately obvious which are the steps to compile the binaries in the current version for each system. 
You should provide a dead simple way for each User to compile these for their system (without having to follow your complicated workflow and recompile the whole application)! That will help most likely help all those linux souls out there, indexing forever...

This is especially relevant, since you are not always in the mood to lend a hand, see:
https://github.com/SublimeCodeIntel/CodeIntelSources/issues/2

sadly, Komodos codebase is mostly a good example for bad design choices and footguns. I for my part will continue to fix their bugs and add features but stay away from any merges and only manually pull in future code that I understand and approve. 

> and as much as I wish I was immortal, I'm not, and if I'm ever not around, I'd want for someone else to know what my workflow was, so anybody can continue from there.
this sounds somewhat mysterious...
